src/imputation/expansion_imputation.py

```diff
@@ -18,38 +18,46 @@ def evaluate_imputed_ixx(
     """
+    imp_class = group["imp_class"].values[0]
+    ExpansionLogger.debug(f"Imputation class: {imp_class}.")
+    ExpansionLogger.debug(f"Master column: {master_col}.")
+
+    # Return the groupby object unaltered if there are no TMI values
+    # in the imputation class
+    TMI_mask = group["imp_marker"] == "TMI"
+    imputed_subgroup = group.loc[TMI_mask]
+
+    if imputed_subgroup.empty:
+        ExpansionLogger.debug(
+            f"Imputation class {imp_class} has no TMI rows to process."
+        )
+        return group
+
     # Make cols into str just in case coming through as ints
     bd_cols = [str(col) for col in break_down_cols]
 
     # Sum the master column, e.g. "211" or "305" for the clear responders
     clear_statuses = ["Clear", "Clear - overridden"]
     clear_mask = group["status"].isin(clear_statuses)
-    sum_master_q = group.loc[clear_mask][master_col].sum()  # scalar
+    sum_master_q = group.loc[clear_mask, master_col].sum()  # scalar
+
+    # Find the imputed value for the master col for the current imputation class
+    # As this value is the same for the whole imputation class, we return the first
+    master_imp_val = group.loc[TMI_mask, f"{master_col}_imputed"].values[0]  # scalar
 
-    # Calculate the imputation columns for the breakdown questions
+    # Calculate the value to enter as the imputed value for the breakdown cols
     for bd_col in bd_cols:
         # Sum the breakdown q for the (clear) responders
-        sum_breakdown_q = group.loc[clear_mask][bd_col].sum()
-
-        # Master col imputed, e.g. 211_imputed or 305_imputed
-        # Sum the breakdown master q imputed for TMI records
-        TMI_mask = group["imp_marker"] == "TMI"
-        master_imp_val = group.loc[TMI_mask][f"{master_col}_imputed"]
-        if master_imp_val.values.any():
-            master_imp_val = master_imp_val.values[0]  # get a single value
-        else:
-            master_imp_val = float("nan")  # assign nan where there are no vals
+        sum_breakdown_q = group.loc[clear_mask, bd_col].sum()
 
-        # Make imputation col equal to original column
-        group[f"{bd_col}_imputed"] = group[bd_col]
+        # calculate the imputed values
+        if master_imp_val > 0:
+            imputed_value = (sum_breakdown_q / sum_master_q) * master_imp_val
+        else:
+            imputed_value = 0
 
-        # Update the imputation column for status encoded 100 and 201
-        # i.e. for non-responders
-        unclear_statuses = ["Form sent out", "Check needed"]
-        unclear_mask = group["status"].isin(unclear_statuses)
-        imputed_value = (sum_breakdown_q / sum_master_q) * master_imp_val
-        # Write imputed value to the non-responder records
-        group.loc[unclear_mask, f"{bd_col}_imputed"] = imputed_value
+        # Update the imputed break-down columns where TMI imputation has been applied
+        group.loc[TMI_mask, f"{bd_col}_imputed"] = imputed_value
```

Comment:
This seems a lot cleaner! Are the 0s we're assigning here wrapped into the question Ana had about do we impute with 0s?

Reply:
This bit is the bug fix- the other question from Ana is something else again.